In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import confusion_matrix, precision_score
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn import svm
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.neural_network import MLPClassifier, MLPRegressor

np.set_printoptions(precision=4)

In [2]:
data = pd.read_csv('final_df.csv')

In [3]:
data.head()

season  round  weather_warm  weather_cold  weather_dry  weather_wet  \
0    2009      1          True         False        False        False   
1    2009      1          True         False        False        False   
2    2009      1          True         False        False        False   
3    2009      1          True         False        False        False   
4    2009      1          True         False        False        False   

   weather_cloudy       driver  grid  podium  ...  constructor_force_india  \
0           False       button     1       1  ...                        0   
1           False  barrichello     2       2  ...                        0   
2           False       vettel     3      13  ...                        0   
3           False       kubica     4      14  ...                        0   
4           False      rosberg     5       6  ...                        0   

   constructor_haas  constructor_lotus_f1  constructor_mclaren  \
0                 0                     0                    0   
1                 0                     0                    0   
2                 0                     0                    0   
3                 0                     0                    0   
4                 0                     0                    0   

   constructor_mercedes  constructor_red_bull  constructor_renault  \
0                     0                     0                    0   
1                     0                     0                    0   
2                     0                     1                    0   
3                     0                     0                    0   
4                     0                     0                    0   

   constructor_sauber  constructor_toro_rosso  constructor_williams  
0                   0                       0                     0  
1                   0                       0                     0  
2                   0                       0                     0  
3                   0                       0                     0  
4                   0                       0                     1  

[5 rows x 61 columns]

## Classification

In [4]:
df = data.copy()
df.podium = df.podium.map(lambda x: 1 if x == 1 else 0)

train = df[df.season <2019]
X_train = train.drop(['driver', 'podium'], axis = 1)
y_train = train.podium

scaler = StandardScaler()
X_train = pd.DataFrame(scaler.fit_transform(X_train), columns = X_train.columns)


In [5]:
def score_classification(model):
    score = 0
    for circuit in df[df.season == 2019]['round'].unique():

        test = df[(df.season == 2019) & (df['round'] == circuit)]
        X_test = test.drop(['driver', 'podium'], axis = 1)
        y_test = test.podium

        #scaling
        X_test = pd.DataFrame(scaler.transform(X_test), columns = X_test.columns)

        # make predictions
        prediction_df = pd.DataFrame(model.predict_proba(X_test), columns = ['proba_0', 'proba_1'])
        prediction_df['actual'] = y_test.reset_index(drop = True)
        prediction_df.sort_values('proba_1', ascending = False, inplace = True)
        prediction_df.reset_index(inplace = True, drop = True)
        prediction_df['predicted'] = prediction_df.index
        prediction_df['predicted'] = prediction_df.predicted.map(lambda x: 1 if x == 0 else 0)

        score += precision_score(prediction_df.actual, prediction_df.predicted)
        
    model_score = score / df[df.season == 2019]['round'].unique().max()
    return model_score

In [6]:
comparison_dict ={'model':[],
                  'params': [],
                  'score': []}

### Logistic Regression

In [7]:
# Logistic Regression

params={'penalty': ['l1', 'l2'],
        'solver': ['saga', 'liblinear'],
        'C': np.logspace(-3,1,20)}

for penalty in params['penalty']:
    for solver in params['solver']:
        for c in params['C']:
            model_params = (penalty, solver, c)
            model = LogisticRegression(penalty = penalty, solver = solver, C = c, max_iter = 10000)
            model.fit(X_train, y_train)
            
            model_score = score_classification(model)
            
            comparison_dict['model'].append('logistic_regression')
            comparison_dict['params'].append(model_params)
            comparison_dict['score'].append(model_score)

### Random Forest Classifier

In [8]:
# Random Forest Classifier

params={'criterion': ['gini', 'entropy'],
        'max_features': [0.8, 'auto', None],
        'max_depth': list(np.linspace(5, 55, 26)) + [None]}

for criterion in params['criterion']:
    for max_features in params['max_features']:
        for max_depth in params['max_depth']:
            model_params = (criterion, max_features, max_depth)
            model = RandomForestClassifier(criterion = criterion, max_features = max_features, max_depth = max_depth)
            model.fit(X_train, y_train)
            
            model_score = score_classification(model)
            
            comparison_dict['model'].append('random_forest_classifier')
            comparison_dict['params'].append(model_params)
            comparison_dict['score'].append(model_score)

### SVM Classifier

In [9]:
# Support Vector Machines

params={'gamma': np.logspace(-4, -1, 20),
        'C': np.logspace(-2, 1, 20),
        'kernel': ['linear', 'poly', 'rbf', 'sigmoid']} 

for gamma in params['gamma']:
    for c in params['C']:
        for kernel in params['kernel']:
            model_params = (gamma, c, kernel)
            model = svm.SVC(probability = True, gamma = gamma, C = c, kernel = kernel )
            model.fit(X_train, y_train)
            
            model_score = score_classification(model)
            
            comparison_dict['model'].append('svm_classifier')
            comparison_dict['params'].append(model_params)
            comparison_dict['score'].append(model_score)

### Neural Network Classifier

In [10]:
# Neural network

params={'hidden_layer_sizes': [(80,20,40,5), (75,25,50,10)], 
        'activation': ['identity', 'logistic', 'tanh', 'relu'], 
        'solver': ['lbfgs'], 
        'alpha': np.logspace(-4,2,20)} 


for hidden_layer_sizes in params['hidden_layer_sizes']:
    for activation in params['activation']:
        for solver in params['solver']:
            for alpha in params['alpha']:
                model_params = (hidden_layer_sizes, activation, solver, alpha )
                model = MLPClassifier(hidden_layer_sizes = hidden_layer_sizes,
                                      activation = activation, solver = solver, alpha = alpha, random_state = 1, max_iter = 100000)
                model.fit(X_train, y_train)

                model_score = score_classification(model)

                comparison_dict['model'].append('neural_network_classifier')
                comparison_dict['params'].append(model_params)
                comparison_dict['score'].append(model_score)

## Regression

In [11]:
df = data.copy()

train = df[df.season <2019]
X_train = train.drop(['driver', 'podium'], axis = 1)
y_train = train.podium

scaler = StandardScaler()
X_train = pd.DataFrame(scaler.fit_transform(X_train), columns = X_train.columns)


In [12]:
def score_regression(model):
    score = 0
    for circuit in df[df.season == 2019]['round'].unique():

        test = df[(df.season == 2019) & (df['round'] == circuit)]
        X_test = test.drop(['driver', 'podium'], axis = 1)
        y_test = test.podium

        #scaling
        X_test = pd.DataFrame(scaler.transform(X_test), columns = X_test.columns)

        # make predictions
        prediction_df = pd.DataFrame(model.predict(X_test), columns = ['results'])
        prediction_df['podium'] = y_test.reset_index(drop = True)
        prediction_df['actual'] = prediction_df.podium.map(lambda x: 1 if x == 1 else 0)
        prediction_df.sort_values('results', ascending = True, inplace = True)
        prediction_df.reset_index(inplace = True, drop = True)
        prediction_df['predicted'] = prediction_df.index
        prediction_df['predicted'] = prediction_df.predicted.map(lambda x: 1 if x == 0 else 0)

        score += precision_score(prediction_df.actual, prediction_df.predicted)

    model_score = score / df[df.season == 2019]['round'].unique().max()
    return model_score

### Linear Regression

In [13]:
# Linear Regression

params={'fit_intercept': ['True', 'False']}

for fit_intercept in params['fit_intercept']:
    model_params = (fit_intercept)
    model = LinearRegression(fit_intercept = fit_intercept)
    model.fit(X_train, y_train)
            
    model_score = score_regression(model)
            
    comparison_dict['model'].append('linear_regression')
    comparison_dict['params'].append(model_params)
    comparison_dict['score'].append(model_score)


### Random Forest Regressor

In [14]:
# Random Forest Regressor

params={'criterion': ['mse'],
        'max_features': [0.8, 'auto', None],
        'max_depth': list(np.linspace(5, 55, 26)) + [None]}

for criterion in params['criterion']:
    for max_features in params['max_features']:
        for max_depth in params['max_depth']:
            model_params = (criterion, max_features, max_depth)
            model = RandomForestRegressor(criterion = criterion,
                                          max_features = max_features, max_depth = max_depth, random_state = 1)
            model.fit(X_train, y_train)
            
            model_score = score_regression(model)
            
            comparison_dict['model'].append('random_forest_regressor')
            comparison_dict['params'].append(model_params)
            comparison_dict['score'].append(model_score)


### SVM Regressor

In [15]:
## Support Vector Machines

params={'gamma': np.logspace(-4, -1, 10),
        'C': np.logspace(-2, 1, 10),
        'kernel': ['linear', 'poly', 'rbf', 'sigmoid']} 

for gamma in params['gamma']:
    for c in params['C']:
        for kernel in params['kernel']:
            model_params = (gamma, c, kernel)
            model = svm.SVR(gamma = gamma, C = c, kernel = kernel)
            model.fit(X_train, y_train)
            
            model_score = score_regression(model)
            
            comparison_dict['model'].append('svm_regressor')
            comparison_dict['params'].append(model_params)
            comparison_dict['score'].append(model_score)


In [76]:
best_models=pd.DataFrame(pd.DataFrame(comparison_dict).groupby('model')['score'].max())
best_models.reset_index(inplace=True)
best_models

model     score
0          linear_regression  0.476190
1        logistic_regression  0.619048
2  neural_network_classifier  0.619048
3               nn_regressor  0.523810
4   random_forest_classifier  0.523810
5    random_forest_regressor  0.428571
6             svm_classifier  0.666667
7              svm_regressor  0.523810

I have chosen the top two scores
1. SVM Classifier = 0.666667
2. Neural Network Classifier = 0.619048

In [83]:
All_models=pd.DataFrame(comparison_dict)


In [81]:
Best_Model_info=pd.merge(best_models,All_models,how='left',left_on=['model','score'],right_on=['model','score'])
Best_Model_info.drop_duplicates(keep=False,inplace=True)

In [82]:
Best_Model_info

model     score  \
0           linear_regression  0.476190   
1           linear_regression  0.476190   
2         logistic_regression  0.619048   
3         logistic_regression  0.619048   
4   neural_network_classifier  0.619048   
5   neural_network_classifier  0.619048   
6                nn_regressor  0.523810   
7                nn_regressor  0.523810   
8                nn_regressor  0.523810   
9                nn_regressor  0.523810   
10               nn_regressor  0.523810   
11               nn_regressor  0.523810   
12               nn_regressor  0.523810   
13               nn_regressor  0.523810   
14               nn_regressor  0.523810   
15               nn_regressor  0.523810   
16               nn_regressor  0.523810   
17               nn_regressor  0.523810   
18               nn_regressor  0.523810   
19               nn_regressor  0.523810   
20               nn_regressor  0.523810   
21               nn_regressor  0.523810   
22               nn_regressor  0.523810   
23               nn_regressor  0.523810   
24               nn_regressor  0.523810   
25   random_forest_classifier  0.523810   
26   random_forest_classifier  0.523810   
27   random_forest_classifier  0.523810   
28    random_forest_regressor  0.428571   
29    random_forest_regressor  0.428571   
30    random_forest_regressor  0.428571   
31             svm_classifier  0.666667   
32              svm_regressor  0.523810   

                                               params  
0                                                True  
1                                               False  
2                 (l1, liblinear, 0.8858667904100823)  
3                      (l2, saga, 0.5455594781168515)  
4   ((80, 20, 40, 5), identity, lbfgs, 11.28837891...  
5   ((75, 25, 50, 10), identity, lbfgs, 11.2883789...  
6            ((80, 20, 40, 5), identity, sgd, 0.0001)  
7   ((80, 20, 40, 5), identity, sgd, 0.00018329807...  
8   ((80, 20, 40, 5), identity, sgd, 0.00033598182...  
9   ((80, 20, 40, 5), identity, sgd, 0.00061584821...  
10  ((80, 20, 40, 5), identity, sgd, 0.00112883789...  
11  ((80, 20, 40, 5), identity, sgd, 0.00206913808...  
12  ((80, 20, 40, 5), identity, sgd, 0.00379269019...  
13  ((80, 20, 40, 5), identity, sgd, 0.00695192796...  
14  ((80, 20, 40, 5), identity, sgd, 0.01274274985...  
15  ((80, 20, 40, 5), identity, sgd, 0.02335721469...  
16  ((80, 20, 40, 5), identity, sgd, 0.04281332398...  
17  ((80, 20, 40, 5), identity, sgd, 0.07847599703...  
18  ((80, 20, 40, 5), identity, sgd, 0.14384498882...  
19  ((80, 20, 40, 5), identity, sgd, 0.26366508987...  
20  ((80, 20, 40, 5), identity, sgd, 0.48329302385...  
21  ((80, 20, 40, 5), identity, sgd, 0.88586679041...  
22  ((80, 20, 40, 5), identity, sgd, 2.97635144163...  
23  ((80, 20, 40, 5), identity, sgd, 5.45559478116...  
24             ((80, 20, 40, 5), identity, sgd, 10.0)  
25                                  (gini, 0.8, 21.0)  
26                                 (gini, None, 47.0)  
27                              (entropy, auto, 51.0)  
28                                   (mse, 0.8, 29.0)  
29                                  (mse, auto, 19.0)  
30                                  (mse, None, 19.0)  
31  (0.0012742749857031334, 0.08858667904100823, s...  
32               (0.001, 4.6415888336127775, sigmoid)

Now that we have hyperparamters of the both models, we will make corresponding models using pytorch.

In [91]:
Best_Model_info.to_csv('Model_Params.csv')